File: go/computer_player.go
package main

import (
	"math"
	"math/rand"
)

// GameState provides context for AI decision making
type GameState struct {
	Round         int
	Players       []PlayerInterface
	ActivePlayers []PlayerInterface
	CurrentLeader PlayerInterface
	CardsInDeck   []*Card
}

type HitOrStayStrategy func(self PlayerInterface, gameState *GameState) bool
type ActionTargetStrategy func(self PlayerInterface, gameState *GameState, actionType ActionType) PlayerInterface

type ComputerPlayer struct {
	BasePlayer
	HitOrStayStrategy            HitOrStayStrategy
	ActionTargetStrategy         ActionTargetStrategy
	PositiveActionTargetStrategy ActionTargetStrategy
}

// NewComputerPlayer creates a new computer player with specified strategy
func NewComputerPlayer(name string, strategy HitOrStayStrategy, actionTargetStrategy ActionTargetStrategy, positiveActionTargetStrategy ActionTargetStrategy) *ComputerPlayer {
	p := &ComputerPlayer{
		HitOrStayStrategy:            strategy,
		ActionTargetStrategy:         actionTargetStrategy,
		PositiveActionTargetStrategy: positiveActionTargetStrategy,
	}

	p.BasePlayer.Init(name)

	return p
}

func (p *ComputerPlayer) GetPlayerIcon() string {
	return "🤖"
}

func (p *ComputerPlayer) MakeHitStayDecision(gameState *GameState) (bool, error) {
	// Always hit if you have a second chance
	if p.HasSecondChance() {
		return true, nil
	}

	return p.HitOrStayStrategy(p, gameState), nil
}

func (p *ComputerPlayer) ChooseActionTarget(gameState *GameState, actionType ActionType) (PlayerInterface, error) {
	return p.ActionTargetStrategy(p, gameState, actionType), nil
}

func (p *ComputerPlayer) ChoosePositiveActionTarget(gameState *GameState, actionType ActionType) (PlayerInterface, error) {
	return p.PositiveActionTargetStrategy(p, gameState, actionType), nil
}

func PlayRoundTo(n int) HitOrStayStrategy {
	return func(self PlayerInterface, gameState *GameState) bool {
		return self.CalculateRoundScore() < n
	}
}

func PlayToBustProbability(p float64) HitOrStayStrategy {
	return func(self PlayerInterface, gameState *GameState) bool {
		return CalculateBustProbability(self, gameState) < p
	}
}

// Helper functions for advanced strategies
func CalculateBustProbability(player PlayerInterface, gameState *GameState) float64 {
	numberCards := make(map[int]bool)
	for _, card := range player.GetHand() {
		if card.Type == NumberCard {
			numberCards[card.Value] = true
		}
	}

	// Count available cards that would cause a bust
	bustCards := 0
	for _, card := range gameState.CardsInDeck {
		if card.Type == NumberCard && numberCards[card.Value] {
			bustCards += 1
		}
	}

	totalCards := len(gameState.CardsInDeck)
	if totalCards == 0 {
		panic("no cards left in deck can't calculate bust probability")
	}

	return float64(bustCards) / float64(totalCards)
}

func HitUntilAheadBy(n int) HitOrStayStrategy {
	return func(self PlayerInterface, gameState *GameState) bool {
		return gameState.CurrentLeader.GetTotalScore()+gameState.CurrentLeader.CalculateRoundScore() < self.GetTotalScore()+self.CalculateRoundScore()+n
	}
}

func AlwaysHitStrategy(self PlayerInterface, gameState *GameState) bool {
	return true
}

func RandomHitOrStayStrategy(self PlayerInterface, gameState *GameState) bool {
	return rand.Intn(2) == 0
}

// Advanced strategies that could beat bust probability < 0.3

// AdaptiveBustProbabilityStrategy adjusts the bust threshold based on game state
func AdaptiveBustProbabilityStrategy(baseProbability float64) HitOrStayStrategy {
	return func(self PlayerInterface, gameState *GameState) bool {
		bustProb := CalculateBustProbability(self, gameState)

		// Adjust threshold based on position in game
		adjustedThreshold := baseProbability

		// More aggressive if behind
		if gameState.CurrentLeader != nil && gameState.CurrentLeader != self {
			leaderScore := gameState.CurrentLeader.GetTotalScore() + gameState.CurrentLeader.CalculateRoundScore()
			myScore := self.GetTotalScore() + self.CalculateRoundScore()
			gap := leaderScore - myScore

			if gap > 50 {
				adjustedThreshold += 0.15 // Much more aggressive when far behind
			} else if gap > 20 {
				adjustedThreshold += 0.1 // More aggressive when behind
			} else if gap < -20 {
				adjustedThreshold -= 0.1 // More conservative when ahead
			}
		}

		// More aggressive late in game
		if gameState.CurrentLeader != nil {
			leaderScore := gameState.CurrentLeader.GetTotalScore()
			if leaderScore > 150 {
				adjustedThreshold += 0.1 // Game is ending soon, take more risks
			}
		}

		return bustProb < adjustedThreshold
	}
}

// ExpectedValueStrategy considers the expected points gained vs risk
func ExpectedValueStrategy(self PlayerInterface, gameState *GameState) bool {
	bustProb := CalculateBustProbability(self, gameState)
	expectedPoints := CalculateExpectedPointsFromHit(self, gameState)
	currentScore := self.CalculateRoundScore()

	// Hit if expected value is positive and risk isn't too high
	expectedValue := expectedPoints * (1 - bustProb)

	// Adjust based on current position
	threshold := 2.0 // Base threshold for expected value
	if gameState.CurrentLeader != nil && gameState.CurrentLeader != self {
		leaderScore := gameState.CurrentLeader.GetTotalScore() + gameState.CurrentLeader.CalculateRoundScore()
		myScore := self.GetTotalScore() + currentScore
		if leaderScore-myScore > 30 {
			threshold = 1.0 // Lower threshold when behind
		}
	}

	return expectedValue > threshold && bustProb < 0.5
}

// HybridStrategy combines multiple factors for decision making
func HybridStrategy(self PlayerInterface, gameState *GameState) bool {
	currentScore := self.CalculateRoundScore()
	bustProb := CalculateBustProbability(self, gameState)

	// Base decision on bust probability
	baseBustThreshold := 0.25

	// Factor 1: Position relative to leader
	if gameState.CurrentLeader != nil {
		leaderScore := gameState.CurrentLeader.GetTotalScore() + gameState.CurrentLeader.CalculateRoundScore()
		myScore := self.GetTotalScore() + currentScore
		gap := leaderScore - myScore

		if gap > 40 {
			baseBustThreshold += 0.2 // Much more aggressive when far behind
		} else if gap > 15 {
			baseBustThreshold += 0.1 // More aggressive when behind
		} else if gap < -15 {
			baseBustThreshold -= 0.1 // More conservative when ahead
		}
	}

	// Factor 2: Round score - be more conservative with high scores
	if currentScore > 30 {
		baseBustThreshold -= 0.1
	} else if currentScore < 10 {
		baseBustThreshold += 0.05
	}

	// Factor 3: Game progress - more aggressive near end
	if gameState.CurrentLeader != nil {
		maxScore := gameState.CurrentLeader.GetTotalScore()
		if maxScore > 160 {
			baseBustThreshold += 0.1
		}
	}

	// Factor 4: Modifier cards - more aggressive with multipliers
	if hasMultiplier(self) {
		baseBustThreshold += 0.05
	}

	return bustProb < baseBustThreshold
}

// GapBasedStrategy focuses on the score gap to other players
func GapBasedStrategy(self PlayerInterface, gameState *GameState) bool {
	if gameState.CurrentLeader == nil {
		return CalculateBustProbability(self, gameState) < 0.3
	}

	leaderScore := gameState.CurrentLeader.GetTotalScore() + gameState.CurrentLeader.CalculateRoundScore()
	myScore := self.GetTotalScore() + self.CalculateRoundScore()
	gap := leaderScore - myScore

	// Dynamic bust threshold based on gap
	var bustThreshold float64
	switch {
	case gap > 60:
		bustThreshold = 0.5 // Very aggressive when far behind
	case gap > 30:
		bustThreshold = 0.4 // Aggressive when behind
	case gap > 10:
		bustThreshold = 0.35 // Moderately aggressive
	case gap > -10:
		bustThreshold = 0.3 // Standard
	case gap > -30:
		bustThreshold = 0.25 // Conservative when ahead
	default:
		bustThreshold = 0.2 // Very conservative when far ahead
	}

	return CalculateBustProbability(self, gameState) < bustThreshold
}

// OptimalStrategy - combines best elements of gap-based and bust probability
func OptimalStrategy(self PlayerInterface, gameState *GameState) bool {
	bustProb := CalculateBustProbability(self, gameState)
	currentScore := self.CalculateRoundScore()

	// Start with dynamic gap-based threshold
	var baseThreshold float64
	if gameState.CurrentLeader != nil {
		leaderScore := gameState.CurrentLeader.GetTotalScore() + gameState.CurrentLeader.CalculateRoundScore()
		myScore := self.GetTotalScore() + currentScore
		gap := leaderScore - myScore

		switch {
		case gap > 50:
			baseThreshold = 0.45 // Very aggressive when far behind
		case gap > 25:
			baseThreshold = 0.35 // Aggressive when behind
		case gap > 10:
			baseThreshold = 0.3 // Moderately aggressive
		case gap > -10:
			baseThreshold = 0.28 // Slightly conservative
		case gap > -25:
			baseThreshold = 0.25 // Conservative when ahead
		default:
			baseThreshold = 0.22 // Very conservative when far ahead
		}
	} else {
		baseThreshold = 0.3 // Default when no leader
	}

	// Adjust for current round score (be more conservative with high scores)
	if currentScore > 35 {
		baseThreshold -= 0.08
	} else if currentScore > 25 {
		baseThreshold -= 0.05
	} else if currentScore < 10 {
		baseThreshold += 0.03
	}

	// Adjust for game state
	if gameState.CurrentLeader != nil {
		maxScore := gameState.CurrentLeader.GetTotalScore()
		if maxScore > 170 {
			baseThreshold += 0.05 // More aggressive near end of game
		} else if maxScore > 150 {
			baseThreshold += 0.03
		}
	}

	// Adjust for modifier cards
	if hasMultiplier(self) && currentScore < 25 {
		baseThreshold += 0.04 // More aggressive with multiplier at low scores
	}

	// Apply minimum and maximum bounds
	if baseThreshold > 0.5 {
		baseThreshold = 0.5
	} else if baseThreshold < 0.15 {
		baseThreshold = 0.15
	}

	return bustProb < baseThreshold
}

func CalculateExpectedPointsFromHit(player PlayerInterface, gameState *GameState) float64 {
	numberCards := make(map[int]bool)
	for _, card := range player.GetHand() {
		if card.Type == NumberCard {
			numberCards[card.Value] = true
		}
	}

	totalPoints := 0.0
	validCards := 0

	for _, card := range gameState.CardsInDeck {
		if card.Type == NumberCard && !numberCards[card.Value] {
			totalPoints += float64(card.Value)
			validCards++
		} else if card.Type == ModifierCard {
			totalPoints += float64(card.GetPoints())
			validCards++
		}
		// Action cards have variable value, approximate as 5 points
		if card.Type == ActionCard {
			totalPoints += 5.0
			validCards++
		}
	}

	if validCards == 0 {
		return 0
	}

	return totalPoints / float64(validCards)
}

func hasMultiplier(player PlayerInterface) bool {
	for _, card := range player.GetHand() {
		if card.Type == ModifierCard && card.Modifier == Multiply2 {
			return true
		}
	}
	return false
}

func TargetLeaderStrategy(self PlayerInterface, gameState *GameState, actionType ActionType) PlayerInterface {
	var leader PlayerInterface
	leaderScore := 0
	for _, player := range gameState.ActivePlayers {
		if actionType == SecondChance && player.HasSecondChance() {
			continue
		}

		if player != self {
			playerScore := player.GetTotalScore() + player.CalculateRoundScore()
			if playerScore > leaderScore {
				leader = player
				leaderScore = playerScore
			}
		}
	}

	// Must target self if no other player is active
	if leader == nil {
		return self
	}

	return leader
}

func TargetLastPlaceStrategy(self PlayerInterface, gameState *GameState, actionType ActionType) PlayerInterface {
	var last PlayerInterface
	lastScore := math.MaxInt
	for _, player := range gameState.ActivePlayers {
		if actionType == SecondChance && player.HasSecondChance() {
			continue
		}

		if player != self {
			playerScore := player.GetTotalScore() + player.CalculateRoundScore()
			if playerScore < lastScore {
				last = player
				lastScore = playerScore
			}
		}
	}

	// Must target self if no other player is active
	if last == nil {
		return self
	}

	return last
}

func TargetRandomStrategy(self PlayerInterface, gameState *GameState, actionType ActionType) PlayerInterface {
	activePlayers := make([]PlayerInterface, 0)
	for _, player := range gameState.Players {
		if actionType == SecondChance && player.HasSecondChance() {
			continue
		}

		if player.IsActive() && player != self {
			activePlayers = append(activePlayers, player)
		}
	}

	// Must target self if no other player is active
	if len(activePlayers) == 0 {
		return self
	}

	return activePlayers[rand.Intn(len(activePlayers))]
}

File: computer_player.py
"""Computer player and AI strategies for Flip 7."""

import math
import random
from typing import Callable, List, Optional

from card import Card, CardType, ActionType, ModifierType
from player import BasePlayer, GameState


HitOrStayStrategy = Callable[[BasePlayer, GameState], bool]
ActionTargetStrategy = Callable[[BasePlayer, GameState, ActionType], BasePlayer]


def calculate_bust_probability(player: BasePlayer, game_state: GameState) -> float:
    """Probability that the next card causes a bust."""
    number_cards = {c.value for c in player.get_hand() if c.type == CardType.NUMBER}
    bust_cards = sum(
        1 for c in game_state.cards_in_deck
        if c.type == CardType.NUMBER and c.value in number_cards
    )
    total = len(game_state.cards_in_deck)
    if total == 0:
        raise RuntimeError("no cards left in deck can't calculate bust probability")
    return bust_cards / total


def calculate_expected_points_from_hit(player: BasePlayer, game_state: GameState) -> float:
    """Expected points from drawing one more card."""
    number_cards = {c.value for c in player.get_hand() if c.type == CardType.NUMBER}
    total_points = 0.0
    valid_cards = 0
    for c in game_state.cards_in_deck:
        if c.type == CardType.NUMBER and c.value not in number_cards:
            total_points += c.value
            valid_cards += 1
        elif c.type == CardType.MODIFIER:
            total_points += c.get_points()
            valid_cards += 1
        elif c.type == CardType.ACTION:
            total_points += 5.0
            valid_cards += 1
    if valid_cards == 0:
        return 0.0
    return total_points / valid_cards


def _has_multiplier(player: BasePlayer) -> bool:
    for c in player.get_hand():
        if c.type == CardType.MODIFIER and c.modifier == ModifierType.MULTIPLY_2:
            return True
    return False


# --- Hit/Stay strategies ---

def play_round_to(n: int) -> HitOrStayStrategy:
    def strategy(self: BasePlayer, game_state: GameState) -> bool:
        return self.calculate_round_score() < n
    return strategy


def play_to_bust_probability(p: float) -> HitOrStayStrategy:
    def strategy(self: BasePlayer, game_state: GameState) -> bool:
        return calculate_bust_probability(self, game_state) < p
    return strategy


def hit_until_ahead_by(n: int) -> HitOrStayStrategy:
    def strategy(self: BasePlayer, game_state: GameState) -> bool:
        leader = game_state.current_leader
        if leader is None:
            return True
        leader_score = leader.get_total_score() + leader.calculate_round_score()
        my_score = self.get_total_score() + self.calculate_round_score()
        return leader_score < my_score + n
    return strategy


def always_hit_strategy(self: BasePlayer, game_state: GameState) -> bool:
    return True


def random_hit_or_stay_strategy(self: BasePlayer, game_state: GameState) -> bool:
    return random.randint(0, 1) == 0


def adaptive_bust_probability_strategy(base_probability: float) -> HitOrStayStrategy:
    def strategy(self: BasePlayer, game_state: GameState) -> bool:
        bust_prob = calculate_bust_probability(self, game_state)
        adjusted = base_probability
        leader = game_state.current_leader
        if leader is not None and leader != self:
            leader_score = leader.get_total_score() + leader.calculate_round_score()
            my_score = self.get_total_score() + self.calculate_round_score()
            gap = leader_score - my_score
            if gap > 50:
                adjusted += 0.15
            elif gap > 20:
                adjusted += 0.1
            elif gap < -20:
                adjusted -= 0.1
        if leader is not None and leader.get_total_score() > 150:
            adjusted += 0.1
        return bust_prob < adjusted
    return strategy


def expected_value_strategy(self: BasePlayer, game_state: GameState) -> bool:
    bust_prob = calculate_bust_probability(self, game_state)
    expected_points = calculate_expected_points_from_hit(self, game_state)
    current_score = self.calculate_round_score()
    expected_value = expected_points * (1 - bust_prob)
    threshold = 2.0
    leader = game_state.current_leader
    if leader is not None and leader != self:
        leader_score = leader.get_total_score() + leader.calculate_round_score()
        my_score = self.get_total_score() + current_score
        if leader_score - my_score > 30:
            threshold = 1.0
    return expected_value > threshold and bust_prob < 0.5


def hybrid_strategy(self: BasePlayer, game_state: GameState) -> bool:
    current_score = self.calculate_round_score()
    bust_prob = calculate_bust_probability(self, game_state)
    base_bust_threshold = 0.25
    leader = game_state.current_leader
    if leader is not None:
        leader_score = leader.get_total_score() + leader.calculate_round_score()
        my_score = self.get_total_score() + current_score
        gap = leader_score - my_score
        if gap > 40:
            base_bust_threshold += 0.2
        elif gap > 15:
            base_bust_threshold += 0.1
        elif gap < -15:
            base_bust_threshold -= 0.1
    if current_score > 30:
        base_bust_threshold -= 0.1
    elif current_score < 10:
        base_bust_threshold += 0.05
    if leader is not None and leader.get_total_score() > 160:
        base_bust_threshold += 0.1
    if _has_multiplier(self):
        base_bust_threshold += 0.05
    return bust_prob < base_bust_threshold


def gap_based_strategy(self: BasePlayer, game_state: GameState) -> bool:
    leader = game_state.current_leader
    if leader is None:
        return calculate_bust_probability(self, game_state) < 0.3
    leader_score = leader.get_total_score() + leader.calculate_round_score()
    my_score = self.get_total_score() + self.calculate_round_score()
    gap = leader_score - my_score
    if gap > 60:
        bust_threshold = 0.5
    elif gap > 30:
        bust_threshold = 0.4
    elif gap > 10:
        bust_threshold = 0.35
    elif gap > -10:
        bust_threshold = 0.3
    elif gap > -30:
        bust_threshold = 0.25
    else:
        bust_threshold = 0.2
    return calculate_bust_probability(self, game_state) < bust_threshold


def optimal_strategy(self: BasePlayer, game_state: GameState) -> bool:
    bust_prob = calculate_bust_probability(self, game_state)
    current_score = self.calculate_round_score()
    leader = game_state.current_leader
    if leader is not None:
        leader_score = leader.get_total_score() + leader.calculate_round_score()
        my_score = self.get_total_score() + current_score
        gap = leader_score - my_score
        if gap > 50:
            base_threshold = 0.45
        elif gap > 25:
            base_threshold = 0.35
        elif gap > 10:
            base_threshold = 0.3
        elif gap > -10:
            base_threshold = 0.28
        elif gap > -25:
            base_threshold = 0.25
        else:
            base_threshold = 0.22
    else:
        base_threshold = 0.3
    if current_score > 35:
        base_threshold -= 0.08
    elif current_score > 25:
        base_threshold -= 0.05
    elif current_score < 10:
        base_threshold += 0.03
    if leader is not None:
        max_score = leader.get_total_score()
        if max_score > 170:
            base_threshold += 0.05
        elif max_score > 150:
            base_threshold += 0.03
    if _has_multiplier(self) and current_score < 25:
        base_threshold += 0.04
    base_threshold = max(0.15, min(0.5, base_threshold))
    return bust_prob < base_threshold


# --- Action target strategies ---

def target_leader_strategy(
    self: BasePlayer, game_state: GameState, action_type: ActionType
) -> BasePlayer:
    leader: Optional[BasePlayer] = None
    leader_score = 0
    for p in game_state.active_players:
        if action_type == ActionType.SECOND_CHANCE and p.has_second_chance():
            continue
        if p != self:
            s = p.get_total_score() + p.calculate_round_score()
            if s > leader_score:
                leader = p
                leader_score = s
    return leader if leader is not None else self


def target_last_place_strategy(
    self: BasePlayer, game_state: GameState, action_type: ActionType
) -> BasePlayer:
    last: Optional[BasePlayer] = None
    last_score: int = 2 ** 31 - 1  # maxint-like
    for p in game_state.active_players:
        if action_type == ActionType.SECOND_CHANCE and p.has_second_chance():
            continue
        if p != self:
            s = p.get_total_score() + p.calculate_round_score()
            if s < last_score:
                last = p
                last_score = s
    return last if last is not None else self


def target_random_strategy(
    self: BasePlayer, game_state: GameState, action_type: ActionType
) -> BasePlayer:
    active: List[BasePlayer] = []
    for p in game_state.players:
        if action_type == ActionType.SECOND_CHANCE and p.has_second_chance():
            continue
        if p.is_active() and p != self:
            active.append(p)
    if not active:
        return self
    return random.choice(active)


class ComputerPlayer(BasePlayer):
    """Computer player that uses configurable hit/stay and action target strategies."""

    def __init__(
        self,
        name: str,
        hit_or_stay_strategy: HitOrStayStrategy,
        action_target_strategy: ActionTargetStrategy,
        positive_action_target_strategy: ActionTargetStrategy,
    ) -> None:
        super().__init__(name)
        self.init(name)
        self._hit_or_stay = hit_or_stay_strategy
        self._action_target = action_target_strategy
        self._positive_action_target = positive_action_target_strategy

    def get_player_icon(self) -> str:
        return "🤖"

    def make_hit_stay_decision(self, game_state: GameState) -> bool:
        if self.has_second_chance():
            return True
        return self._hit_or_stay(self, game_state)

    def choose_action_target(
        self, game_state: GameState, action_type: ActionType
    ) -> BasePlayer:
        return self._action_target(self, game_state, action_type)

    def choose_positive_action_target(
        self, game_state: GameState, action_type: ActionType
    ) -> BasePlayer:
        return self._positive_action_target(self, game_state, action_type)
